README.md
Ruby/ZOOM
=========

Ruby/ZOOM provides a Ruby binding to the Z39.50 Object-Orientation Model 
(ZOOM), an abstract object-oriented programming interface to a subset of the 
services specified by the Z39.50 standard, also known as the international 
standard ISO 23950.

This software is based (and therefore depends) on YAZ, a free-software 
implementation of the Z39.50/SRW/SRU standards, but could be easily ported to 
any ZOOM compliant implementation.

Ruby/ZOOM is free-software, covered by the GNU Lesser General Public License 
and provided without any warranties of any kind.

Requirements
------------

    Ruby:       http://www.ruby-lang.org/
    YAZ (*):    http://www.indexdata.dk/yaz/ 

    * : Make sure you did pass the --enabled-shared option to the configure
        script before building YAZ.  Ruby/ZOOM requires a YAZ shared library 
        and YAZ does not build it by default. 

Install
-------

    0. gem install alexandria-zoom

Build from Source using bundler
-----------------

    0. add to your Gemfile:
    gem "alexandria-zoom"
    1. bundle

Samples
-------

    Some programming examples are available in the `sample' directory. 

Canonical Sample Program
------------------------

  To give a flavour of the Ruby binding, here is ZOOM's equivalent of the
  ``Hello World'' program: a tiny Z39.50 client that fetches and displays
  the MARC record for Farlow & Brett Surman's The Complete Dinosaur from
  the Library of Congress.


    require 'zoom'

    ZOOM::Connection.open('z3950.loc.gov', 7090) do |conn|
      conn.database_name = 'Voyager'
      conn.preferred_record_syntax = 'USMARC'
      rset = conn.search('@attr 1=7 0253333490')
      p rset[0]
    end

ZOOM Extended Services
----------------------

  With release 0.4.1 extended services support has been added which allows
  create, update and delete of records. This only works for XML.
  Here's an example adapted from the test suite. 

    require 'zoom'

    new_record = File.read('programming_ruby.xml')

    ZOOM::Connection.open('localhost:99999/test') do |conn|
      p = conn.package
      p.wait_action = 'waitIfPossible'
      p.action = 'specialUpdate'
      p.record = new_record
      p.send('update')
      p.send('commit')     
    end


Copying
-------

    Copyright (c) 2005 Laurent Sansonetti <lrz@chopine.be>
      
    This library is free software.
    You can distribute/modify this program under the terms of
    the GNU LESSER GENERAL PUBLIC LICENSE Version 2.1.

Project Website
---------------

    http://ruby-zoom.rubyforge.org


ChangeLog
v0.6.0 Fri Apr 28 21:56:11 CEST 2023
  - Support Ruby 2.7 through 3.2
  - Use pkg-config instead of yaz-config to get compiler and linker flags for YAZ
  - Created forked gem 'alexandria-zoom'
  - Update gem infrastructure
  - Silence warnings regarding allocator functions for T_DATA classes on Ruby 3.2

v0.5.0 Sat Aug 15 11:39:00 EST 2015
  - Added support for Ruby 1.9.2.
  - Gemspec and build tooling improvements.

v0.4.1 Tue Nov 20 08:05:53 EST 2007 
  - record retrieval has changed with batch_download branch merged into trunk.
    this allows for records to be downloaded in a batch and tries to fall back
    to record at a time download if batch download fails. 
  - removed second definition of databaseName from rbzoompackage.c
  - added RDoc style documentation to C source
  - added tests for batch retrieval and a connection within a Thread

v0.4.0 - checking in support of ZOOM::Connection#package
  - creating a .4 version of gem
  - added Walter McGinnis, Nicolai Moles-Benfell To Authors
  - src/rbzoomoptions.c: Added support for rubyfying option names that 
    contain hyphens or full stops.
  - src/rbzoomconnection.c: Added ZOOM::Connection#package, a factory method 
    that returns an instance of a ZOOM::Package object.
  - added src/rbzoompackage.c:

v0.3.0  Tue Jul 10 11:57:40 EDT 2007
  - new format for this file, all changes are summarized under a version
    no need to recreate commit messages in a text file when svn can tell us them
  - added Rakefile
  - applied patch from Jason Ronallo to enable raw method on records
  - added test suite
  - created gem for rubyforge

2007-05-05  Ed Summers <ehs@pobox.com>
  - added Ed Summers
  - src/lib/* removed marc.rb since it isn't used and conflicts with the marc 
    gem

# Older format preserved for posterity ...

2006-09-21  Laurent Sansonetti  <lrz@chopine.be>

  * ChangeLog,
    README,
    src/**/*.{c,h,rb}:

  s/gnome.org/chopine.be/g.

2006-09-21  Devin Bayer  <devin@freeshell.org>

  * .cvsignore:

  Ignores zoom.bundle.

   * sample/needle.rb:

  Updated the needle demo application.

   * src/rbzoomconnection.c:

  Do not fail when reading options.

2006-08-22  Laurent Sansonetti  <lrz@chopine.be>

  * AUTHORS:

  - Updated my e-mail address ;
  - Added Devin.
 
2006-08-22  Devin Bayer  <devin@freeshell.org>

  * src/rbzoomrecord.c:

  Added ZOOM::Record#raw, to grab a marc record in marc format.

2006-08-20  Devin Bayer  <devin@freeshell.org>

  * sample/needle.rb:

  Added new sample, contributed by Devin Bayer.

=== Ruby/ZOOM 0.2.2 ===

2005-11-08  Laurent Sansonetti  <lrz@chopine.be>

  * README:

  Added a note about YAZ's --enable-shared problem.

  * AUTHORS:

  Added Pascal and Matt.

2005-11-08  Matt Vanderpol  <matt@vanderpol.net>

  * src/rbzoomquery.c (.new_prefix, .new_cql):

  Fixed constructors.

=== Ruby/ZOOM 0.2.1 ===

2005-09-23  Pascal Terjan  <pterjan@linuxfr.org>

  * src/rbzoomoptions.c:

  Fixes segfault when rubyfying the option name.

=== Ruby/ZOOM 0.2.0 ===

2005-08-02  Laurent Sansonetti  <lrz@chopine.be>

  * src/rbzoomresultset.rb (#[]):

  Retrieves the record one by one using ZOOM_resultset_record instead
  of getting them all in once with ZOOM_resultset_records (for a strange
  reason sometimes the resultset was not empty but ZOOM_resultset_records
  used to return empty records).

2005-07-19  Laurent Sansonetti  <lrz@chopine.be>

  * src/lib/marc.rb:

  Added an experimental MARC decoder.

=== Ruby/ZOOM 0.1.0 ===

2005-03-18  Laurent Sansonetti  <lrz@chopine.be>

  * COPYING.LIB,
    AUTHORS,
    README:

  Added.

  * src/rbzoomconnection.c,
    src/rbzoomquery.c:

  Fixed documentation.

  * src/rbzoomresultset.c (#[]):

  - Range objects are supported ;
  - Do not clone nil objects.

2005-03-17  Laurent Sansonetti  <lrz@chopine.be>

  * doc/.cvsignore,
    doc/gendoc.sh:

  Added.

  * src/rbzoom.c,
    src/rbzoom.h,
    src/rbzoomconnection.c,
    src/rbzoomoptions.c,
    src/rbzoomquery.c,
    src/rbzoomrecord.c,
    src/rbzoomresultset.c:

  Added extdoc documentation + LGPL headers.

2005-03-16  Laurent Sansonetti  <lrz@chopine.be>

  * src/rbzoomrecord.c (#database, #syntax, #render, #xml):

  Added.

  * src/rbzoom.c,
    sample/hello.rb:

  Renamed the module name as 'ZOOM'.

2005-03-14  Laurent Sansonetti  <lrz@chopine.be>

  Initial RubyForge revision.


ext/zoom/rbzoomquery.c
/*
 * Copyright (C) 2005 Laurent Sansonetti <lrz@chopine.be>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "rbzoom.h"

#ifdef MAKING_RDOC_HAPPY
mZoom = rb_define_module("ZOOM");
#endif

/* Class: ZOOM::Query
 * Search queries.
 */
static VALUE cZoomQuery;

static VALUE
rbz_query_make (ZOOM_query query)
{
    return query != NULL
        ? Data_Wrap_Struct (cZoomQuery,
                            NULL,
                            ZOOM_query_destroy,
                            query)
        : Qnil;
}

ZOOM_query
rbz_query_get (VALUE obj)
{
    ZOOM_query query;
        
    Data_Get_Struct (obj, struct ZOOM_query_p, query);
    assert (query != NULL);

    return query;
}

/*
 * call-seq: new_prefix(prefix)
 *
 * prefix: PQF notation.
 *
 * Creates a RPN query using the given PQF notation.
 *
 * Returns: a newly created ZOOM::Query object.
 */
static VALUE
rbz_query_new_prefix (VALUE self, VALUE prefix)
{
    ZOOM_query query;

    query = ZOOM_query_create ();
    ZOOM_query_prefix (query, RVAL2CSTR (prefix)); 
    
    return rbz_query_make (query);
}

/* call-seq:
 * 	 new_cql(prefix)
 *
 * prefix: CQL notation.
 *
 * Creates a CQL query using the given CQL notation.
 *
 * Returns: a newly created ZOOM::Query object.
 */
static VALUE
rbz_query_new_cql (VALUE self, VALUE cql)
{
    ZOOM_query query;

    query = ZOOM_query_create ();
    ZOOM_query_cql (query, RVAL2CSTR (cql)); 
    
    return rbz_query_make (query);
}

/*
 * call-seq: new_sort_by(criteria)
 *
 * criteria: a sort criteria.
 *
 * Creates a sort query from the YAZ sorting notation.
 *
 * Returns: a newly created ZOOM::Query object.
 */
static VALUE
rbz_query_new_sort_by (VALUE self, VALUE criteria)
{
    ZOOM_query query;

    query = ZOOM_query_create ();
    ZOOM_query_sortby (rbz_query_get (self), RVAL2CSTR (criteria)); 
    
    return rbz_query_make (query);
}

void
Init_zoom_query (VALUE mZoom)
{
    VALUE c;

    c = rb_define_class_under (mZoom, "Query", rb_cObject); 
    rb_undef_alloc_func(c);

    rb_define_singleton_method (c, "new_prefix", rbz_query_new_prefix, 1);
    rb_define_singleton_method (c, "new_cql", rbz_query_new_cql, 1);
    rb_define_singleton_method (c, "new_sort_by", rbz_query_new_sort_by, 1);
            
    cZoomQuery = c;
}


ext/zoom/rbzoomrecord.c
/*
 * Copyright (C) 2005 Laurent Sansonetti <lrz@chopine.be>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "rbzoom.h"

#ifdef MAKING_RDOC_HAPPY
mZoom = rb_define_module("ZOOM");
#endif


/* 
 * A record object is a retrieval record on the client side - created from 
 * result sets.
 */
static VALUE cZoomRecord;

VALUE
rbz_record_make (ZOOM_record record)
{
    return record != NULL
        ? Data_Wrap_Struct (cZoomRecord,
                            NULL,
                            ZOOM_record_destroy,
                            record)
        : Qnil;
}

static ZOOM_record
rbz_record_get (VALUE obj)
{
    ZOOM_record record;
        
    Data_Get_Struct (obj, struct ZOOM_record_p, record);
    assert (record != NULL);

    return record;
}

static char _type [128];

static const char *
rbz_record_type (const char *form, int argc, VALUE *argv)
{
    VALUE charset_from;
    VALUE charset_to;

    if (argc == 0)
        return form;

    rb_scan_args (argc, argv, "11", &charset_from, &charset_to);
   
    memset (_type, 0, sizeof _type);
    
    if (NIL_P (charset_to))
        snprintf (_type, sizeof _type, "%s; charset=%s", form, 
                  RVAL2CSTR (charset_from));
    else
        snprintf (_type, sizeof _type, "%s; charset=%s,%s", form, 
                  RVAL2CSTR (charset_from), RVAL2CSTR (charset_to));

    return _type;
}

/*
 * call-seq: 
 * 	database(charset_from=nil, charset_to=nil)
 *
 * charset_from: the name of the charset to convert from (optional).
 *
 * charset_to: the name of the charset to convert to (optional).
 *
 * Returns: the database name of the record.
 */
static VALUE
rbz_record_database (int argc, VALUE *argv, VALUE self)
{
    return CSTR2RVAL (ZOOM_record_get (rbz_record_get (self),
                                       rbz_record_type ("database", argc, argv),
                                       NULL));    
}

/*
 * call-seq: 
 * 	syntax(charset_from=nil, charset_to=nil)
 *
 * charset_from: the name of the charset to convert from (optional).
 *
 * charset_to: the name of the charset to convert to (optional).
 *
 * Returns: the symbolic transfer syntax name of the record.
 */
static VALUE
rbz_record_syntax (int argc, VALUE *argv, VALUE self)
{
    return CSTR2RVAL (ZOOM_record_get (rbz_record_get (self),
                                       rbz_record_type ("syntax", argc, argv),
                                       NULL));    
}

/*
 * call-seq: 
 * 	render(charset_from=nil, charset_to=nil)
 *
 * charset_from: the name of the charset to convert from (optional).
 *
 * charset_to: the name of the charset to convert to (optional).
 *
 * Returns: a display friendly description of the record.
 */
static VALUE
rbz_record_render (int argc, VALUE *argv, VALUE self)
{
    return CSTR2RVAL (ZOOM_record_get (rbz_record_get (self),
                                       rbz_record_type ("render", argc, argv),
                                       NULL));    
}

/*
 * call-seq: 
 * 	xml(charset_from=nil, charset_to=nil)
 *
 * charset_from: the name of the charset to convert from (optional).
 *
 * charset_to: the name of the charset to convert to (optional).
 *
 * Returns an XML description of the record.  SRW/SRU and Z39.50 records with 
 * transfer syntax XML are returned verbatim.  MARC records are returned in 
 * MARCXML (converted from ISO2709 to MARCXML by YAZ). GRS-1 and OPAC records are 
 * not supported for this form. 
 * 
 * Returns: an XML description of the record.
 */
static VALUE
rbz_record_xml (int argc, VALUE *argv, VALUE self)
{
    return CSTR2RVAL (ZOOM_record_get (rbz_record_get (self),
                                       rbz_record_type ("xml", argc, argv),
                                       NULL));    
}

/*
 * call-seq: 
 * 	raw
 *
 * MARC records are returned in ISO2709.
 * GRS-1 and OPAC records are not supported for this form. 
 * 
 * Returns: an ISO2709 record.
 */

static VALUE
rbz_record_raw (int argc, VALUE *argv, VALUE self)
{
    return CSTR2RVAL (ZOOM_record_get (rbz_record_get (self),
                                       rbz_record_type ("raw", argc, argv),
                                       NULL));    
}

void
Init_zoom_record (VALUE mZoom)
{
    VALUE c;
    
    c = rb_define_class_under (mZoom, "Record", rb_cObject); 
    rb_undef_alloc_func(c);

    rb_undef_method (CLASS_OF (c), "new");

    rb_define_method (c, "database", rbz_record_database, -1);
    rb_define_method (c, "syntax", rbz_record_syntax, -1);
    rb_define_method (c, "render", rbz_record_render, -1);
    rb_define_alias (c, "to_s", "render");
    rb_define_method (c, "xml", rbz_record_xml, -1);
    rb_define_method (c, "raw", rbz_record_raw, -1);
    
    cZoomRecord = c;
}


ext/zoom/rbzoompackage.c
/*
 * Copyright (C) 2007 Katipo Communications, Ltd.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "rbzoom.h"
#include <stdio.h>

#ifdef MAKING_RDOC_HAPPY
mZoom = rb_define_module("ZOOM");
#endif

/* Document-class: ZOOM::Package
 *
 * This class represents an Extended Services Package: an instruction to the server 
 * to do something not covered by the core parts of the Z39.50 standard 
 */
static VALUE cZoomPackage;


static ZOOM_package
rbz_package_get (VALUE obj)
{
    ZOOM_package package;

	if (cZoomPackage == Qnil)
		rb_raise(rb_eRuntimeError, "cZoomPackage is nil: has destroy() already been called on this Package?");
        
    Data_Get_Struct (obj, struct ZOOM_package_p, package);
    assert (package != NULL);

    return package;
}


/*
 * call-seq: 
 * 	make(connection, options)
*
*  Creates a ZOOM::Package from the connection and options specified.
*
*  Returns: the created ZOOM::Package or Qnil.
*/
VALUE
rbz_package_make (ZOOM_connection connection, ZOOM_options options)
{

  ZOOM_package package;

  package =  ZOOM_connection_package(connection, options);

  if (cZoomPackage == Qnil)
	rb_raise(rb_eRuntimeError, "cZoomPackage is nil: has destroy() already been called on this Package?");

  return package != NULL
      ? Data_Wrap_Struct (cZoomPackage,
                          NULL,
                          ZOOM_package_destroy,
                          package)
      : Qnil;
}


/*
 * call-seq:
 * 	set_option(key, value)
 *
 * key: the name of the option, as a string.
 *
 * value: the value of this option (as a string, integer or boolean).
 *
 * Sets an option on the package.
 * 
 * Returns: self.
 */
static VALUE
rbz_package_set_option (VALUE self, VALUE key, VALUE val)
{   

	ZOOM_package package;
    
    package = rbz_package_get (self);
    ZOOM_package_option_set (package,
                                RVAL2CSTR (key),
                                RVAL2CSTR (rb_obj_as_string (val)));
   
    return self;
}


/*
 * call-seq:
 * 	get_option(key)
 *
 * key: the name of the option, as a string.
 *
 * Gets the value of a package's option.
 * 
 * Returns: the value of the given option, as a string, integer or boolean.
 */
static VALUE
rbz_package_get_option (VALUE self, VALUE key)
{
	ZOOM_package package;
    const char *value;
 
    package = rbz_package_get (self);
 
    value = ZOOM_package_option_get (package,
                                        RVAL2CSTR (key));

    return zoom_option_value_to_ruby_value (value);
}

/*
 * call-seq:
 * 	send(type)
 *
 * type:  the actual extended service package type to be sent, as a string.
 *
 * Sends the package.
 * 
 * Returns: self.
 */
static VALUE
rbz_package_send(VALUE self, VALUE type)
{
    ZOOM_package package;
	const char *typeChar;

    package = rbz_package_get (self);

    typeChar = StringValuePtr(type);
    ZOOM_package_send(package, typeChar);
  
    return self;
}



/* Interface to a subset of the Z39.50 extended services.
*/
void
Init_zoom_package (VALUE mZoom)
{
    VALUE c;
    
    c = rb_define_class_under (mZoom, "Package", rb_cObject); 
    rb_undef_alloc_func(c);

	/* Remove the default constructor to force initialization through Connection#package. */
    rb_undef_method (CLASS_OF (c), "new");

	/* Instance methods */
    rb_define_method (c, "set_option", rbz_package_set_option, 2);
    rb_define_method (c, "get_option", rbz_package_get_option, 1);
    rb_define_method (c, "send", rbz_package_send, 1);

	// Common Options
    define_zoom_option (c, "package-name");
	define_zoom_option (c, "user-id");
	define_zoom_option (c, "function");
	define_zoom_option (c, "waitAction");
	define_zoom_option (c, "targetReference");
	
	// Item Order, type must be set to itemorder in ZOOM_package_send.
	define_zoom_option (c, "contact-name");
	define_zoom_option (c, "contact-phone");
	define_zoom_option (c, "contact-email");
	define_zoom_option (c, "itemorder-item");
	
	// Record Update, type must be set to update in ZOOM_package_send.
	define_zoom_option (c, "action");
	define_zoom_option (c, "recordIdOpaque");
	define_zoom_option (c, "recordIdNumber");
	define_zoom_option (c, "record");
	define_zoom_option (c, "syntax");
	define_zoom_option (c, "databaseName");
	define_zoom_option (c, "correlationInfo.note");
	define_zoom_option (c, "correlationInfo.id");
	define_zoom_option (c, "elementSetName");
	
	// Database Create, type must be set to create in ZOOM_package_send.
	// Database Drop, type must be set to drop in ZOOM_package_send.
	    
    cZoomPackage = c;
}


ext/zoom/rbzoomconnection.c
/*
 * Copyright (C) 2005 Laurent Sansonetti <lrz@chopine.be>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "rbzoom.h"

#ifdef MAKING_RDOC_HAPPY
mZoom = rb_define_module("ZOOM");
#endif


/* Document-class: ZOOM::Connection
 * The Connection object is a session with a target.
 */
static VALUE cZoomConnection;


static VALUE
rbz_connection_make (ZOOM_connection connection)
{
    return connection != NULL
        ? Data_Wrap_Struct (cZoomConnection,
                            NULL,
                            ZOOM_connection_destroy,
                            connection)
        : Qnil;
}

static ZOOM_connection
rbz_connection_get (VALUE obj)
{
    ZOOM_connection connection;
        
    Data_Get_Struct (obj, struct ZOOM_connection_p, connection);
    assert (connection != NULL);

    return connection;
}

#define RAISE_IF_FAILED(connection)                     \
    do {                                                \
        int error;                                      \
        const char *errmsg;                             \
        const char *addinfo;                            \
                                                        \
        error = ZOOM_connection_error (connection,      \
                                       &errmsg,         \
                                       &addinfo);       \
        if (error != 0)                                 \
            rb_raise (rb_eRuntimeError, "%s (%d) %s",   \
                      errmsg, error, addinfo);          \
    }                                                   \
    while (0)

void rbz_connection_check(VALUE obj)
{
	ZOOM_connection connection;
	
	connection = rbz_connection_get(obj);
  	RAISE_IF_FAILED (connection);
}


/*
 * call-seq: 
 * 	open(host, port=nil) { |conn| ... }
 *
 * host: hostname of the target to connect to.
 *
 * port: network port of the target to connect to.
 *
 * A convenience method that creates a new connection and attempts to 
 * establish a network connection to the given target, basically calling
 * ZOOM::Connection.new and ZOOM::Connection#connect.
 *  
 * If a block is given, then it will be called once the connection is 
 * established, passing a reference to the connection object as a parameter, 
 * and destroying the connection automatically at the end of the block.
 * With no block, this method just returns the connection object.
 *
 * Returns: a newly created ZOOM::Connection object.
 */
static VALUE
rbz_connection_open (int argc, VALUE *argv, VALUE self)
{
    VALUE host;
    VALUE port;
    ZOOM_connection connection;
    VALUE rb_connection;
    
    rb_scan_args (argc, argv, "11", &host, &port);

    connection = ZOOM_connection_new (RVAL2CSTR (host),
                                      NIL_P (port) ? 0 : FIX2INT (port));
    RAISE_IF_FAILED (connection);
    
    rb_connection = rbz_connection_make (connection);
    if (rb_block_given_p ()) {
        rb_yield(rb_connection);
        return Qnil;
    }
    return rb_connection;
}

/*
 * call-seq: new(options=nil)
 *
 * options: options for the connection, as a Hash object.
 *
 * Creates a new connection object, but does not establish a network connection
 * immediately, allowing you to specify options before (if given). You can 
 * thus establish the connection using ZOOM::Connection#connect.
 * 
 * Returns: a newly created ZOOM::Connection object.
 */
static VALUE
rbz_connection_new (int argc, VALUE *argv, VALUE self)
{
    ZOOM_options options;
    ZOOM_connection connection;
    VALUE rb_options;
    
    rb_scan_args (argc, argv, "01", &rb_options);

    if (NIL_P (rb_options))
        options = ZOOM_options_create ();
    else
        options = ruby_hash_to_zoom_options (rb_options);

    connection = ZOOM_connection_create (options);
    ZOOM_options_destroy (options);
    RAISE_IF_FAILED (connection);

    return rbz_connection_make (connection);
}

/*
 * call-seq:
 * 	connect(host, port=nil)
 *
 * host: hostname of the target to connect to.
 *
 * port: network port of the target to connect to.
 *
 * Establishes a network connection to the target specified by the given
 * arguments.  If no port is given, 210 will be used.  A colon in the host
 * string denotes the beginning of a port number.  If the host string includes
 * a slash, the following part specifies a database for the connection.
 *
 * You can also prefix the host string with a scheme followed by a colon.
 * The default scheme is tcp (Z39.50 protocol).  The scheme http selects SRW
 * over HTTP.
 *
 * This method raises an exception on error.
 *
 * Returns: self.
 */
static VALUE
rbz_connection_connect (int argc, VALUE *argv, VALUE self)
{
    ZOOM_connection connection;
    VALUE host;
    VALUE port;
    
    rb_scan_args (argc, argv, "11", &host, &port);
  
    connection = rbz_connection_get (self);
    ZOOM_connection_connect (connection, 
                             RVAL2CSTR (host), 
                             NIL_P (port) ? 0 : FIX2INT (port));
    RAISE_IF_FAILED (connection); 

    return self;
}

/*
 * call-seq:
 * 	set_option(key, value)
 *
 * key: the name of the option, as a string.
 *
 * value: the value of this option (as a string, integer or boolean).
 *
 * Sets an option on the connection.
 * 
 * Returns: self.
 */
static VALUE
rbz_connection_set_option (VALUE self, VALUE key, VALUE val)
{
    ZOOM_connection connection;
    
    connection = rbz_connection_get (self);
    ZOOM_connection_option_set (connection,
                                RVAL2CSTR (key),
                                RVAL2CSTR (rb_obj_as_string (val)));
    RAISE_IF_FAILED (connection); 
    
    return self;
}

/*
 * call-seq: 
 * 	get_option(key)
 *
 * key: the name of the option, as a string.
 *
 * Gets the value of a connection's option.
 * 
 * Returns: the value of the given option, as a string, integer or boolean.
 */
static VALUE
rbz_connection_get_option (VALUE self, VALUE key)
{
    ZOOM_connection connection;
    const char *value;
 
    connection = rbz_connection_get (self);
    value = ZOOM_connection_option_get (connection,
                                        RVAL2CSTR (key));

    return zoom_option_value_to_ruby_value (value);
}

/*
 * call-seq: 
 * 	search(criterion)
 *
 * criterion: the search criterion, either as a ZOOM::Query object or as a string,
 * representing a PQF query.
 *  
 * Searches the connection from the given criterion.  You can either create and
 * pass a reference to a ZOOM::Query object, or you can simply pass a string
 * that represents a PQF query.
 *
 * This method raises an exception on error.
 * 
 * Returns: a result set from the search, as a ZOOM::ResultSet object,
 * empty if no results were found.
 */
static VALUE
rbz_connection_search (VALUE self, VALUE criterion)
{
    ZOOM_connection connection;
    ZOOM_resultset resultset;

    connection = rbz_connection_get (self);
    if (TYPE (criterion) == T_STRING)
        resultset = ZOOM_connection_search_pqf (connection,
                                                RVAL2CSTR (criterion));
    else
        resultset = ZOOM_connection_search (connection,
                                            rbz_query_get (criterion));
    RAISE_IF_FAILED (connection); 
    assert (resultset != NULL);
  
    return rbz_resultset_make (resultset);
}


/*
 * Constructs a new extended services ZOOM::Package using this connections host information.
 *
 * Note: The Perl script passes this connections options if already set, otherwise constructs a new ZOOM::Option object. 
 * Currently this method always constructs a new ZOOM::Option object for each package.
 * 
 * Returns: a new ZOOM::Package object.
 */
static VALUE
rbz_connection_package(VALUE self)
{
  ZOOM_connection connection;
  ZOOM_options options;
  VALUE package;

  connection = rbz_connection_get (self);
  options = ZOOM_options_create ();
  package = rbz_package_make(connection, options);
  return package;
}


void
Init_zoom_connection (VALUE mZoom)
{
    VALUE c;

    c = rb_define_class_under (mZoom, "Connection", rb_cObject); 
    rb_undef_alloc_func(c);

    rb_define_singleton_method (c, "open", rbz_connection_open, -1);
    rb_define_singleton_method (c, "new", rbz_connection_new, -1);
    rb_define_method (c, "connect", rbz_connection_connect, -1);
    rb_define_method (c, "set_option", rbz_connection_set_option, 2);
    rb_define_method (c, "get_option", rbz_connection_get_option, 1);
    rb_define_method (c, "package", rbz_connection_package, 0);

    define_zoom_option (c, "implementationName");
    define_zoom_option (c, "user");
    define_zoom_option (c, "group");
    define_zoom_option (c, "password");
    define_zoom_option (c, "host");
    define_zoom_option (c, "proxy");
    define_zoom_option (c, "async");
    define_zoom_option (c, "maximumRecordSize");
    define_zoom_option (c, "preferredMessageSize");
    define_zoom_option (c, "lang");
    define_zoom_option (c, "charset");
    define_zoom_option (c, "serverImplementationId");
    define_zoom_option (c, "targetImplementationName");
    define_zoom_option (c, "serverImplementationVersion");
    define_zoom_option (c, "databaseName");
    define_zoom_option (c, "piggyback");
    define_zoom_option (c, "smallSetUpperBound");
    define_zoom_option (c, "largeSetLowerBound");
    define_zoom_option (c, "mediumSetPresentNumber");
    define_zoom_option (c, "smallSetElementSetName");
    define_zoom_option (c, "mediumSetElementSetName");

    /* herited from Zoom::ResultSet */
    define_zoom_option (c, "start");
    define_zoom_option (c, "count");
    define_zoom_option (c, "presentChunk");
    define_zoom_option (c, "elementSetName");
    define_zoom_option (c, "preferredRecordSyntax");
    define_zoom_option (c, "schema");
    define_zoom_option (c, "setname");
    
    rb_define_method (c, "search", rbz_connection_search, 1);
    
    cZoomConnection = c;
}
